### PR TITLE
Console UI updates

### PR DESCRIPTION
**Pull request by afshin**
cf. https://github.com/jupyter/jupyterlab/issues/648

cc: @ellisonbg

### CODE REVIEW

**Comment by ellisonbg** [comment]
OK I just spent some time playing with this, focusing on the interaction/UX of it. Observations:

* It is pretty painful to have to manually adjust the location of the splitter to show the code. You either have to make the input area panel too large (to show the maximum lines you might enter) or it will be too small some of the time. When it is too small, the scrolling behavior is not fun.
* The side-by-side mode isn't useful on a laptop, but I could imagine it might be on larger screens with longer input. But awkward to have the input area on the right.
* I played around with the chrome/firefox consoles and both automatically increase the vertical size of the input area as you add more lines.
* The main difference between the chrome/firefox consoles is:
  - Firefox always pins the active input area to the bottom
  - Chrome only pins the active area to the bottom once you have filled the scroll area
  - In practice, the chrome/firefox consoles act identically once you have typed a few lines.

Because of these observations, I propose:

* Follow the firefox model of pinning the input to the bottom of the console and auto-expanding the panel to fit the code cell always (probably still want a max-height that will cause it to scroll, but it can be very large).
* Reconsider the side-by-side mode. Maybe move the input to the left? Maybe keep the split panel for that?

**Comment by ellisonbg** [comment]
Trying out the new behavior, but noticing that a plain enter now always runs code, even if the input is not complete. Regression?

**Comment by ellisonbg** [comment]
Awesome, thanks for working on this...

I think I like the flexbox based approach here and pinning the active cell to the bottom. However, a bit tough to try out:

* As described enter always runs code, even if incomplete
* The input div doesn't quite fully expand to fit the active cell

screen shot 2016-09-07 at 12 58 29 pm

File: src/console/content.ts [hunk 1]
```diff
@@ -65,38 +65,63 @@ import {
...
 export
-class ConsoleWidget extends Widget {
+class ConsoleContent extends Widget {
   /**
    * Construct a console widget.
    */
-  constructor(options: ConsoleWidget.IOptions) {
-    super();
+  constructor(options: ConsoleContent.IOptions) {
+    super({ node: document.createElement('div') });
```
Comment (blink1073) [inline src/console/content.ts]: This is already the default behavior of a Widget.

File: src/console/content.ts [hunk 2]
```diff
@@ -65,38 +65,63 @@ import {
...
     this.addClass(CONSOLE_CLASS);
 
-    let layout = new PanelLayout();
+    // Create the panels that holds the content and input.
```
Comment (blink1073) [inline src/console/content.ts]: hold*

File: src/console/content.ts [hunk 3]
```diff
@@ -65,38 +65,63 @@ import {
...
+    this._content = new Panel();
+    this._input = new Panel();
+    this._content.addClass(CONTENT_CLASS);
+    this._input.addClass(INPUT_CLASS);
+
+    // Insert the content and input panes into the widget.
+    this.node.appendChild(this._content.node);
```
Comment (blink1073) [inline src/console/content.ts]: Why aren't we using a panel layout for this?

File: src/console/content.ts [hunk 4]
```diff
@@ -65,38 +65,63 @@ import {
...
    */
-  constructor(options: ConsoleWidget.IOptions) {
-    super();
+  constructor(options: ConsoleContent.IOptions) {
+    super({ node: document.createElement('div') });
     this.addClass(CONSOLE_CLASS);
 
-    let layout = new PanelLayout();
+    // Create the panels that holds the content and input.
+    this._content = new Panel();
```
Comment (blink1073) [inline src/console/content.ts]: Do these still need to be panels?

File: src/console/content.ts [hunk 5]
```diff
@@ -428,46 +450,57 @@ class ConsoleWidget extends Widget {
...
     /**
-     * The mime renderer for the console widget.
+     * The orientation of the console content.
+     *
+     * #### Notes
+     * This setting indicates how the prompt and output are separated. The
+     * default value is `'vertical'`.
      */
-    rendermime: IRenderMime;
+    orientation?: 'horizontal' | 'vertical';
```
Comment (blink1073) [inline src/console/content.ts]: Not used anymore, correct?

File: src/console/content.ts (hunk 4 above)
Comment (afshin) [inline src/console/content.ts]: Yes, they have widgets added to them.

File: src/console/content.ts (hunk 3 above)
Comment (afshin) [inline src/console/content.ts]: We could ... is that better than simply adding them to the DOM?

File: src/console/content.ts (hunk 5 above)
Comment (afshin) [inline src/console/content.ts]: Good call.

File: src/console/content.ts (hunk 3 above)
Comment (blink1073) [inline src/console/content.ts]: We get the layout semantics/messages vs a simple node append.

Comment (afshin) [inline src/console/content.ts]: Hm, that's fair.

**Comment by blink1073** [comment]
LGTM, I'll let @ellisonbg merge.

File: examples/console/index.css [hunk 6]
```diff
@@ -19,6 +19,12 @@ body {
   bottom: 4px;
 }
 
+
+.p-CommandPalette {
```
Comment (ellisonbg) [inline examples/console/index.css]: Hmmm shouldn't this be somewhere else?

**Comment by ellisonbg** [comment]
One minor thing to fix:

The previous cells div no longer scrolls, overflow should be set to auto on that one.

Once that is done we can merge and iterate...

**Comment by ellisonbg** [comment]
Ahh, I see where the problem is: there is a bit of css in the base.less that should be moved over to index.less. That will fix the overflow issue, but may also cause other problems - I think the base.css is stale...

**Comment by ellisonbg** [comment]
One other thing, can you put the active cell inside a wrapper div for some styling purposes?

File: examples/console/index.css (hunk 6 above)
Comment (afshin) [inline examples/console/index.css]: This is just the console example application's command palette, so it's in the style sheet for that example.

**Comment by afshin** [comment]
@ellisonbg,

> One other thing, can you put the active cell inside a wrapper div for some styling purposes?

Which one do you mean? The current prompt? Its immediate parent is:

```css
.jp-ConsoleContent-input {}
```

And it is itself targetable:
```css
.jp-ConsoleContent-prompt {}
```

Other than that, this should be good to go (and hopefully before I'm working tomorrow morning you can merge it because it's blocking the other code I'm working on). Cheers!

**Comment by ellisonbg** [comment]
Do, missed that!! Looks good, merging! I will work a bit more on the styling of this tomorrow. Totally awesome!

**Pull request merged**